File: android/src/main/kotlin/com/haishinkit/haishin_kit/RtmpStreamHandler.kt
package com.haishinkit.haishin_kit

import android.content.Context
import android.hardware.camera2.CameraCharacteristics
import android.media.MediaFormat.KEY_LEVEL
import android.media.MediaFormat.KEY_PROFILE
import android.os.Handler
import android.util.Size
import android.view.WindowManager
import com.haishinkit.codec.CodecOption
import com.haishinkit.event.Event
import com.haishinkit.event.IEventListener
import com.haishinkit.haishinkit.ProfileLevel
import com.haishinkit.media.AudioRecordSource
import com.haishinkit.media.Camera2Source
import com.haishinkit.rtmp.RtmpStream
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.lang.Exception

class RtmpStreamHandler(
    private val plugin: HaishinKitPlugin, handler: RtmpConnectionHandler?
) : MethodChannel.MethodCallHandler, IEventListener, EventChannel.StreamHandler {
    companion object {
        private const val TAG = "RtmpStream"
    }

    private var instance: RtmpStream? = null
        set(value) {
            field?.dispose()
            field = value
        }
    private var channel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private var camera: Camera2Source? = null
        set(value) {
            field?.close()
            field = value
        }

    init {
        handler?.instance?.let {
            instance = RtmpStream(it)
        }
        channel = EventChannel(
            plugin.flutterPluginBinding.binaryMessenger, "com.haishinkit.eventchannel/${hashCode()}"
        )
        channel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "$TAG#getHasAudio" -> {
                result.success(instance?.audioSetting?.muted)
            }

            "$TAG#setHasAudio" -> {
                val value = call.argument<Boolean?>("value")
                value?.let {
                    instance?.audioSetting?.muted = !it
                }
                result.success(null)
            }

            "$TAG#getHasVideo" -> {
                result.success(null)
            }

            "$TAG#setHasVideo" -> {
                result.success(null)
            }

            "$TAG#setFrameRate" -> {
                val value = call.argument<Int?>("value")
                value?.let {
                    instance?.videoSetting?.frameRate = it
                }
                result.success(null)
            }

            "$TAG#setSessionPreset" -> {
                // for iOS
                result.success(null)
            }

            "$TAG#setAudioSettings" -> {
                val source = call.argument<Map<String, Any?>>("settings") ?: return
                (source["bitrate"] as? Int)?.let {
                    instance?.audioSetting?.bitRate = it
                }
                result.success(null)
            }

            "$TAG#setVideoSettings" -> {
                val source = call.argument<Map<String, Any?>>("settings") ?: return
                (source["width"] as? Int)?.let {
                    instance?.videoSetting?.width = it
                }
                (source["height"] as? Int)?.let {
                    instance?.videoSetting?.height = it
                }
                (source["frameInterval"] as? Int)?.let {
                    instance?.videoSetting?.IFrameInterval = it
                }
                (source["bitrate"] as? Int)?.let {
                    instance?.videoSetting?.bitRate = it
                }
                (source["profileLevel"] as? String)?.let {
                    try {
                        val profileLevel = ProfileLevel.valueOf(it)
                        var options = mutableListOf<CodecOption>()
                        options.add(CodecOption(KEY_PROFILE, profileLevel.profile))
                        options.add(CodecOption(KEY_LEVEL, profileLevel.level))
                        instance?.videoSetting?.options = options
                    } catch (ignored: Exception) {
                        // Do nothing, use default setting
                    }
                }
                result.success(null)
            }

            "$TAG#attachAudio" -> {
                val source = call.argument<Map<String, Any?>>("source")
                if (source == null) {
                    instance?.attachAudio(null)
                } else {
                    instance?.attachAudio(AudioRecordSource(plugin.flutterPluginBinding.applicationContext))
                }
                result.success(null)
            }

            "$TAG#attachVideo" -> {
                val source = call.argument<Map<String, Any?>>("source")
                if (source == null) {
                    instance?.attachVideo(null)
                    camera = null
                } else {
                    var facing = 0
                    when (source["position"]) {
                        "front" -> {
                            facing = CameraCharacteristics.LENS_FACING_FRONT
                        }

                        "back" -> {
                            facing = CameraCharacteristics.LENS_FACING_BACK
                        }
                    }
                    camera = Camera2Source(plugin.flutterPluginBinding.applicationContext)
                    camera?.let {
                        instance?.attachVideo(camera)
                    }
                    val handler = Handler()
                    handler.postDelayed({
                        if (instance?.drawable != null) {
                            camera?.open(facing)
                        }
                    }, 750)
                }
                result.success(null)
            }

            "$TAG#registerTexture" -> {
                val netStream = instance
                if (netStream?.drawable == null) {
                    val texture = NetStreamDrawableTexture(plugin.flutterPluginBinding)
                    texture.attachStream(netStream)
                    if (camera?.stream != null) {
                        camera?.open()
                    }
                    result.success(texture.id)
                } else {
                    val texture = (netStream.drawable as? NetStreamDrawableTexture)
                    val width = call.argument<Double>("width") ?: 0
                    val height = call.argument<Double>("height") ?: 0
                    texture?.imageExtent = Size(width.toInt(), height.toInt())
                    (plugin.flutterPluginBinding.applicationContext.getSystemService(Context.WINDOW_SERVICE) as? WindowManager)?.defaultDisplay?.orientation?.let {
                        netStream.deviceOrientation = it
                    }
                    result.success(texture?.id)
                }
            }

            "$TAG#publish" -> {
                instance?.publish(call.argument("name"))
                result.success(null)
            }

            "$TAG#play" -> {
                val name = call.argument<String>("name")
                if (name != null) {
                    instance?.play(name)
                }
                result.success(null)
            }

            "$TAG#close" -> {
                instance?.close()
                result.success(null)
            }

            "$TAG#dispose" -> {
                eventSink?.endOfStream()
                instance?.close()
                camera?.close()
                instance?.dispose()
                instance = null
                camera = null
                plugin.onDispose(hashCode())
                result.success(null)
            }
        }
    }

    override fun handleEvent(event: Event) {
        val map = HashMap<String, Any?>()
        map["type"] = event.type
        map["data"] = event.data
        plugin.uiThreadHandler.post {
            eventSink?.success(map)
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
    }
}

File: ios/Classes/RTMPStreamHandler.swift
import Foundation
import Flutter
import HaishinKit
import AVFoundation
import VideoToolbox

final class RTMPStreamHandler: NSObject {
    private let plugin: HaishinKitPlugin
    private var texture: HKStreamFlutterTexture?
    private var instance: RTMPStream?
    private var eventSink: FlutterEventSink?
    private var eventChannel: FlutterEventChannel?
    private var subscription: Task<(), Error>?

    init(plugin: HaishinKitPlugin, handler: RTMPConnectionHandler) {
        self.plugin = plugin
        super.init()
        let id = Int(bitPattern: ObjectIdentifier(self))
        if let registrar = plugin.registrar {
            self.eventChannel = FlutterEventChannel(name: "com.haishinkit.eventchannel/\(id)", binaryMessenger: registrar.messenger())
            self.eventChannel?.setStreamHandler(self)
        } else {
            self.eventChannel = nil
        }
        if let connection = handler.instance {
            let instance = RTMPStream(connection: connection)
            plugin.mixer?.addOutput(instance)
            self.instance = instance
        }
    }
}

extension RTMPStreamHandler: MethodCallHandler {
    // MARK: MethodCallHandler
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            guard
                let arguments = call.arguments as? [String: Any?] else {
                result(nil)
                return
            }
            switch call.method {
            case
                "RtmpStream#getHasAudio",
                "RtmpStream#setHasAudio",
                "RtmpStream#getHasVudio",
                "RtmpStream#setHasVudio",
                "RtmpStream#setFrameRate",
                "RtmpStream#setSessionPreset",
                "RtmpStream#attachAudio",
                "RtmpStream#attachVideo",
                "RtmpStream#setScreenSettings":
                plugin.mixer?.handle(call, result: result)
            case "RtmpStream#setAudioSettings":
                guard
                    let settings = arguments["settings"] as? [String: Any?] else {
                    result(nil)
                    return
                }
                if let bitrate = settings["bitrate"] as? NSNumber {
                    var audioSettings = await instance?.audioSettings ?? .default
                    audioSettings.bitRate = bitrate.intValue
                    await instance?.setAudioSettings(audioSettings)
                }
                result(nil)
            case "RtmpStream#setVideoSettings":
                guard
                    let settings = arguments["settings"] as? [String: Any?] else {
                    result(nil)
                    return
                }
                var videoSettings = await instance?.videoSettings ?? .default
                if let bitrate = settings["bitrate"] as? NSNumber {
                    videoSettings.bitRate = bitrate.intValue
                }
                if let width = settings["width"] as? NSNumber, let height = settings["height"] as? NSNumber {
                    videoSettings.videoSize = CGSize(width: .init(width.intValue), height: .init(height.intValue))
                }
                if let frameInterval = settings["frameInterval"] as? NSNumber {
                    videoSettings.maxKeyFrameIntervalDuration = frameInterval.int32Value
                }
                if let profileLevel = settings["profileLevel"] as? String {
                    videoSettings.profileLevel = ProfileLevel(rawValue: profileLevel)?.kVTProfileLevel ?? ProfileLevel.H264_Baseline_AutoLevel.kVTProfileLevel
                }
                await instance?.setVideoSettings(videoSettings)
                result(nil)
            case "RtmpStream#play":
                _ = try? await instance?.play(arguments["name"] as? String)
                result(nil)
            case "RtmpStream#publish":
                _ = try? await instance?.publish(arguments["name"] as? String)
                result(nil)
            case "RtmpStream#registerTexture":
                guard
                    let registry = plugin.registrar?.textures() else {
                    result(nil)
                    return
                }
                if let texture {
                    result(texture.id)
                } else {
                    let texture = HKStreamFlutterTexture(registry: registry)
                    self.texture = texture
                    await instance?.addOutput(texture)
                    result(texture.id)
                }
            case "RtmpStream#unregisterTexture":
                guard
                    let registry = plugin.registrar?.textures() else {
                    result(nil)
                    return
                }
                if let textureId = arguments["id"] as? Int64 {
                    registry.unregisterTexture(textureId)
                }
                result(nil)
            case "RtmpStream#updateTextureSize":
                guard
                    (plugin.registrar?.textures()) != nil else {
                    result(nil)
                    return
                }
                if let texture {
                    if let width = arguments["width"] as? NSNumber,
                       let height = arguments["height"] as? NSNumber {
                        texture.bounds = CGSize(width: width.doubleValue, height: height.doubleValue)
                    }
                    result(texture.id)
                } else {
                    result(nil)
                }
            case "RtmpStream#close":
                _ = try? await instance?.close()
                result(nil)
            case "RtmpStream#dispose":
                if let instance {
                    plugin.mixer?.removeOutput(instance)
                }
                _ = try? await instance?.close()
                instance = nil
                plugin.onDispose(id: Int(bitPattern: ObjectIdentifier(self)))
                result(nil)
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }
}

extension RTMPStreamHandler: FlutterStreamHandler {
    // MARK: FlutterStreamHandler
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
}
